### ISSUE TITLE
Library order is still not correct

### ISSUE THREAD

**Issue by stanhu** [pkgconf/pkgconf#322]
@kaniini This was supposedly fixed in pkgconf v2.0 (#268), but I'm still seeing an issue. Attached is the reproduction step with the `.pc` files I'm trying to use:

```shell
tar xvf test-pc.tar.gz
PKG_CONFIG_PATH=pkgconfig pkg-config --libs re2
```

[test-pc.tar.gz](https://github.com/pkgconf/pkgconf/files/12573655/test-pc.tar.gz)

On Windows, with `pkgconf` v2.0.3 I see the incorrect order. For example, notice that `-labsl_spinlock_wait` precedes `-labsl_flags`, but in v1.8.0 `-labsl_spinlock_wait` is towards the end as it should be:

```
-pthread -L/home/stanhu/github/re2/ports/x64-mingw-ucrt/x86_64-w64-mingw32/libre2/2023-09-01/lib -lr
e2 -L/home/stanhu/github/re2/ports/x64-mingw-ucrt/x86_64-w64-mingw32/abseil/20230802.0/lib -labsl_sp
inlock_wait -labsl_flags -labsl_bad_optional_access -labsl_string_view -labsl_strings_internal -labs
l_flags_program_name -labsl_graphcycles_internal -labsl_kernel_timeout_internal -labsl_civil_time -l
absl_time_zone -labsl_symbolize -ldbghelp -labsl_debugging_internal -labsl_demangle_internal -labsl_
malloc_internal -labsl_flags_internal -labsl_flags_marshalling -labsl_log_severity -labsl_str_format
_internal -labsl_flags_reflection -labsl_flags_private_handle_accessor -labsl_flags_commandlineflag
-labsl_flags_commandlineflag_internal -labsl_flags_config -labsl_cord -labsl_crc32c -labsl_crc_inter
nal -labsl_crc_cpu_detect -labsl_cordz_functions -labsl_cordz_handle -labsl_stacktrace -labsl_cord_i
nternal -labsl_cordz_info -labsl_crc_cord_state -labsl_city -labsl_bad_variant_access -labsl_low_lev
el_hash -labsl_int128 -labsl_hash -labsl_hashtablez_sampler -labsl_exponential_biased -labsl_time -l
absl_raw_logging_internal -labsl_throw_delegate -labsl_base -ladvapi32 -labsl_raw_hash_set -labsl_st
rings -labsl_synchronization
```

However, with `pkgconf` v1.8.0, I see the correct order:

```
-pthread -L/home/stanhu/github/re2/ports/x64-mingw-ucrt/x86_64-w64-mingw32/libre2/2023-09-01/lib -lr
e2 -L/home/stanhu/github/re2/ports/x64-mingw-ucrt/x86_64-w64-mingw32/abseil/20230802.0/lib -labsl_fl
ags -labsl_flags_internal -labsl_flags_marshalling -labsl_flags_reflection -labsl_flags_private_hand
le_accessor -labsl_flags_commandlineflag -labsl_flags_commandlineflag_internal -labsl_flags_config -
labsl_flags_program_name -labsl_cord -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -
labsl_cordz_handle -labsl_crc_cord_state -labsl_crc32c -labsl_crc_internal -labsl_crc_cpu_detect -la
bsl_raw_hash_set -labsl_hash -labsl_city -labsl_bad_variant_access -labsl_low_level_hash -labsl_hash
tablez_sampler -labsl_exponential_biased -labsl_bad_optional_access -labsl_str_format_internal -labs
l_synchronization -labsl_graphcycles_internal -labsl_kernel_timeout_internal -labsl_stacktrace -labs
l_symbolize -ldbghelp -labsl_debugging_internal -labsl_demangle_internal -labsl_malloc_internal -lab
sl_time -labsl_civil_time -labsl_strings -labsl_string_view -labsl_strings_internal -labsl_base -lad
vapi32 -labsl_spinlock_wait -labsl_int128 -labsl_throw_delegate -labsl_raw_logging_internal -labsl_l
og_severity -labsl_time_zone
```

**Comment by bruchar1** [comment]
I'm facing a similar issue. It causes me an access violation in protobuf initialisation when running one of my testsuite in debug. Using pkg-config 0.28, or downgrading to pkgconf < 1.9 solves the problem.

**Comment by LigH-de** [comment]
This issue disables the whole [media-autobuild suite](https://github.com/m-ab-s/media-autobuild_suite) compiling a bunch of A/V related libraries in MSYS2/MinGW with GCC or CLang and linking them together to a custom FFmpeg. Being unable to link libbrotli is one of the first issues we noticed; not having updated the suite for a while and trying a workaround for it, I noticed that this is not the only problem in my case, fontconfig could not link bzip2 in my installation with very similar errors.

**Comment by chewi** [comment]
I'm guessing this is the same issue that was reported in [libxml2](https://gitlab.gnome.org/GNOME/libxml2/-/issues/602). In this case, `-llzma -lz` needs to come after `-lxml2`. Reverting 90b65c866bbb63fb1b867d11d7ff0b180855c65e fixes it, but then many of the arguments get duplicated.

1.8.1:
```
-L/local32/lib -lbluray -L/tmp/xml2/../../lib -lxml2 -liconv -lm -lws2_32 -llzma -pthread -lpthread -lz
```

2.0.3:
```
-L/local32/lib -lbluray -llzma -pthread -lpthread -lz -L/tmp/xml2/../../lib -lxml2 -liconv -lm -lws2_32
```

90b65c866bbb63fb1b867d11d7ff0b180855c65e reverted:
```
-L/local32/lib -lbluray -L/tmp/xml2/../../lib -lxml2 -liconv -lm -lws2_32 -llzma -pthread -lpthread -lz -L/tmp/xml2/../../lib -lxml2 -liconv -lm -lws2_32
```

**Comment by chewi** [comment]
Hmm I tried the fix in #327, but it didn't help in this case.

**Comment by dg0yt** [comment]
#327 is for a different problem: It about constructing the correct order of the directories where to search for pc files.
(But the search order has impact on the inital list of libs, and so it has impact on the final list of libs.)

I started #327 with adding tests (running them in the GH actions of my fork.) I think this might also work for this issue.

(And maybe this issue might explain other problems which prevent a pkgconf upgrade in vcpkg.)

**Comment by kaniini** [comment]
This is because the new solver tries to cancel dependency cycles ahead of time (e.g. when it creates the flattened solution from the resolved DAG) rather than at evaluation time.

**Comment by dg0yt** [comment]
> This is because the new solver tries to cancel dependency cycles ahead of time (e.g. when it creates the flattened solution from the resolved DAG) rather than at evaluation time.

Hm, I think the concern isn't the "Why", but the "Does it work correctly".

Assuming a `<MAIN_LIB> <LINK_LIBS>` input pattern, every lib in `<LINK_LIBS>` must still occur after `<MAIN_LIB>` in the flattened result in the general case (*no cycles*).
https://github.com/pkgconf/pkgconf/issues/322#issuecomment-1751986169 reported that this is not the case for MAIN_LIB `-libxml2`, LINK_LIBS `-llzma -lz`.

In vcpkg, we have to ports with complex dependencies which still fail when trying to upgrade pkgconf (graphviz, ffmpeg). Initially I though it is a CMake problem. At least in graphviz I could restore the order of link libraries with additonal CMake dependencies. But ffmpeg doesn't even use CMake. So probably the root cause is this issue.

**Comment by kaniini** [comment]
Sometimes I add comments to bug reports as a reminder to myself what the cause is, so that I can begin thinking about a solution to the problem.  As I spend 100+ hours a week maintaining free software, there are many issues on my radar at any given time, hence the solution of adding triage notes to reported issues.

**Comment by kaniini** [comment]
By the way, pkg-config format does not have any concept of main lib vs link libs, except to use Requires/Requires.private.

Instead the Libs/Libs.private are just bare tokens which get appended as .pc files are evaluated.  pkgconf tries to be smart and do a mergeback for fragments, but in this case the underlying solution is causing fragments to be missed.

**Comment by dg0yt** [comment]
Maybe it is necessary to make it clear when a comment is "a reminder to self". The comment is read by people which may be in a very similar situation. Having seen three failed attempts to upgrade to `pkgconf` to 2.0 in vcpkg, and in the light of the combined effect of #322 and #327, I was concerned that the impact of the current bugs in 2.0+ isn't fully understood.

> By the way, pkg-config format does not have any concept of main lib vs link libs, except to use Requires/Requires.private.

The explicit concepts are clear to me. In a strict sense, this would imply that pkgconf can't remove anything from the expanded initial `Libs` because common Unix linker expect link libraries in a particular order, and pc file authors write `Libs[.private]` in that order.

> Instead the Libs/Libs.private are just bare tokens which get appended as .pc files are evaluated. pkgconf tries to be smart and do a mergeback for fragments

Basically I just made explicit the assumption which can be the ground for this behaviour: If a lib appears as first item in `Libs` (of any pc file during expansion), every following lib must remain after it. Such as `-llzma -lz` need to remain after `-llibxml2`.

**Comment by kaniini** [comment]
Yes, 90b65c8 generally limited the expansion depth to `2` for `--cflags` and `--libs`.  But this should be fine as the flattened solution, should, in theory be equivalent to allowing infinite recursion at that point.

The fact that it does *not* is due to some edges being cancelled in the flattening process, as I said earlier.  This is a solver bug, and the correct place to fix it is the solver.

**Comment by kaniini** [comment]
> Basically I just made explicit the assumption which can be the ground for this behaviour: If a lib appears as first item in Libs (of any pc file during expansion), every following lib must remain after it. Such as -llzma -lz need to remain after -llibxml2.

pkgconf does retain the order in this case, where the dependencies are spelled out in `Libs`.

The reproducible test-case is when `Requires` is used, and references the module.  In this case, `liblzma` and `zlib` modules get placed in the wrong spot on the flattened solution.

**Comment by dg0yt** [comment]
> The reproducible test-case is when `Requires` is used, and references the module. In this case, `liblzma` and `zlib` modules get placed in the wrong spot on the flattened solution.

Yes, I have the reproducer now:
~~~
Name: depth-0
Description: test
Version: 1.0
Requires.private: private-libs-duplication
~~~
puts `-lprivate` after the `Requires`. i.e. the result is different from private-libs-duplication although it shouldn't.

**Comment by dg0yt** [comment]
`Requires: bar` (with `--static`) is even simpler:
~~~
$ PKG_CONFIG_PATH=../tests/lib1 ./pkgconf depth-0 --static --libs
-L/test/lib -lfoo -L/test/lib -lbar
$ PKG_CONFIG_PATH=../tests/lib1 ./pkgconf bar --static --libs
-L/test/lib -lbar -lfoo
~~~

**Comment by kaniini** [comment]
We can use `--digraph` with graphviz to visualize the solutions with and without recursion:

`pkgconf --with-path=tests/lib1 --static --digraph depth-0 | dot -Tpng -oimg.png`
![img](https://github.com/pkgconf/pkgconf/assets/1522444/19b7eaa1-9e10-4674-8463-3b08f2d66f9f)

`pkgconf --with-path=tests/lib1 depth-0 --static --digraph --maximum-traverse-depth=2 | dot -Tpng -oimg.png`
![img](https://github.com/pkgconf/pkgconf/assets/1522444/ed329e8a-e38d-41b3-946b-c744ea8f2517)

**Comment by dg0yt** [comment]
Reproducer:
https://github.com/pkgconf/pkgconf/compare/master...dg0yt:pkgconf:linklibs?expand=1

**Comment by kaniini** [comment]
I have incorporated that commit into the testsuite.  Visualizing the solution is really interesting... when expansion is enabled, `bar` winds up as an orphaned dependency:

`pkgconf --with-path=tests/lib1 --static --digraph --maximum-traverse-depth=10 metapackage-3 | dot -Tpng -oimg.png`
![img](https://github.com/pkgconf/pkgconf/assets/1522444/9d82aed0-9518-407c-84b6-1ac8d0e3165e)

When only the flattened solution is enabled, `foo` becomes part of the flattened solution, but `bar` does not.  Instead, `bar` becomes connected to `foo` without being in the flattened solution directly:

`pkgconf --with-path=tests/lib1 --static --digraph --maximum-traverse-depth=2 metapackage-3 | dot -Tpng -oimg.png`
![img](https://github.com/pkgconf/pkgconf/assets/1522444/f5dd3d6f-3e3b-4b15-834a-850396ec29c6)

This is, needless to say, very odd.

**Comment by kaniini** [comment]
Nevermind, that was a bug in `--digraph`.  Fixing that bug, we get:

![img](https://github.com/pkgconf/pkgconf/assets/1522444/3665520e-04fa-4987-962c-f2e07f3a6c56)

So I think the problem is that when we pull in `foo` via `requires.private`, we don't add *it* to the solution as a `requires.private` solution, since normally it is a `requires`.  I think, instead, what we want to do is add transitive edges pulled into the DAG via `requires.private` as `requires.private` edges themselves.

And that will fix this bug.

**Comment by kaniini** [comment]
Getting a little bit closer with 752a982:

![img](https://github.com/pkgconf/pkgconf/assets/1522444/1a2f0964-6f74-450a-9ab2-805318e40b53)

But we still need to demote that `bar -> foo` edge.  Or maybe just delete it in the flattened solution.

**Comment by dg0yt** [comment]
Is there any need to treat `x.private` different from  `x` when `--static` is given?
(Of course `Requires.private` should come after `Libs.private`.)

**Comment by kaniini** [comment]
There is not, and indeed, the solution I went with is to just explicitly cancel all edges that have been flattened.  And that mostly works, although `-lfoo` and `-lbar` are now backwards.

**Comment by kaniini** [comment]
Ok, sorted the `-lfoo` and `-lbar` issue, it was a sorting problem in the flattener.  With that fix in place, there are just a handful of testsuite failures to examine, most of which are in the more ... exotic ... tests (for example: testing how infinite loops should be broken), so that could either be problematic or totally fine.

**Comment by kaniini** [comment]
While investigating the test failures, I found an interesting edge case that needs to be rectified, tests/lib1/private-libs-duplication.pc is not being solved correctly by the flattener:

![img](https://github.com/pkgconf/pkgconf/assets/1522444/c4da7e71-0787-4b2e-8966-3bdaa156bb7c)

Will work on this more tomorrow.

**Comment by dg0yt** [comment]
I wasn't able to reproduce the same pictures with `master`...

I tried to find out more, but I still have only fuzzy insight. AFAICT there are two issues:
- The breaking of circular dependency is also triggering in non-circular cases. This is why some edges are not added. 
  This can be avoided by tracking ancestors via markers instead of same-run via serials, https://github.com/pkgconf/pkgconf/commit/8edfc4cea6142b4991fd927be5df32228de249ff.
- The immediate dependencies of virtual:world are treated differently from deep deep dependencies. This leads to some extra edges. (Affecting metapackage-3.)
  This changes when separating input and output virtual:world, https://github.com/pkgconf/pkgconf/commit/462c118bb411bfc132218585f614d552d6df2364.

The two commits work for the meson builds. I can't check why metapackage-3 still fails for the autotools builds.
https://github.com/dg0yt/pkgconf/actions/runs/6523638100

`master`:
![meta-before](https://github.com/pkgconf/pkgconf/assets/13567791/868b6568-94bc-467f-b1be-5db9dbc58982)

With fix for circular dep handling:
![meta-intermediate](https://github.com/pkgconf/pkgconf/assets/13567791/a119e0a1-392f-4bfc-bab4-15fb0dc84c4b)

Plus fix for immediate deps of virtual:world:
![meta-after](https://github.com/pkgconf/pkgconf/assets/13567791/f7dbe3c5-cae9-4bf8-84e5-cbd1a3115a97)

For completeness, this is the final graph for private-libs-duplication:
![missing-after](https://github.com/pkgconf/pkgconf/assets/13567791/f79e79a8-c740-4665-82c7-d04eb0c3312b)

**Comment by dg0yt** [comment]
@kaniini Is there anything else I can do to help move this forward? This is really important for use cases where a working pkgconf is most useful.

**Comment by dg0yt** [comment]
The two proposed commits on top of master seem to solve the graphviz and ffmpeg regressions in vcpkg.

**Comment by dg0yt** [comment]
@kaniini Is there anything else I can do to help move this forward? This is really a blocker for use cases where a working pkgconf is most useful. 
And my "setup costs" increase when I have to come back to this after weeks.

**Comment by kaniini** [comment]
I’ll merge the patches in the morning.  Sorry, have been really busy lately.

**Comment by kaniini** [comment]
Should be fixed with 2.1.0 and the ancestor tracking work by @dg0yt.

**Comment by dg0yt** [comment]
Thanks @kaniini. I'm giving 2.1.0 a run in https://github.com/microsoft/vcpkg/pull/34228.

**Comment by WinterMute** [comment]
> Should be fixed with 2.1.0 and the ancestor tracking work by @dg0yt.

This is still broken for me

On linux/macOS I get this output for the devkitPro ppc freetype package. freetype-config calls pkg-config to get libraries.

`/opt/devkitpro/portlibs/ppc/bin/freetype-config --libs 
-L/opt/devkitpro/portlibs/ppc/lib -lfreetype -lbz2 -lpng16 -lz -lm -lz -lharfbuzz -lm`

On windows, under msys2 bash
`$ /opt/devkitpro/portlibs/ppc/bin/freetype-config --libs
-L/opt/devkitpro/portlibs/ppc/lib -lpng16 -lm -lz -lharfbuzz -lm -L/opt/devkitpro/portlibs/ppc/lib -L/opt/devkitpro/portlibs/ppc/lib -lz -lfreetype -lbz2
`

You can see the library order gets reversed.

pkgconf 2.1.0 is installed.

`$ pacman -Ss pkg-config
msys/pkgconf 2.1.0-1 [installed]
    pkg-config compatible utility which does not depend on glib
`

It may also be worth mentioning that there is a circular dependency between freetype2 & harfbuzz

Is there anything I can do to help figure this out? Should I open a new issue?

**Comment by dg0yt** [comment]
I would suggest to open a new issue when you can provide a reproducible example directly with `pkgconf`.
And cycles don't help.
